core/unix/coredump.c

```diff
@@ -0,0 +1,362 @@
+        }
+        // Write the section names section.
+        if (os_write(elf_file, (void *)string_table, string_table_offset) !=
+            string_table_offset) {
+            os_close(elf_file);
+            return false;
+        }
+        memquery_iterator_stop(&iter);
+    }
+
+    if (memquery_iterator_start(&iter, NULL, /*may_alloc=*/false)) {
+        // TODO i#7046: Handle multiple program headers.
+        ELF_OFF file_offset = sizeof(ELF_HEADER_TYPE) +
+            1 /*program_header_count*/ * sizeof(ELF_PROGRAM_HEADER_TYPE);
+        while (memquery_iterator_next(&iter)) {
+            // Skip non-readable section.
+            if (iter.prot == MEMPROT_NONE || strcmp(iter.comment, VVAR_SECTION) == 0) {
+                continue;
+            }
+            ELF_WORD flags = SHF_ALLOC | SHF_MERGE;
+            if (iter.prot & PROT_WRITE) {
+                flags |= SHF_WRITE;
+            }
+            ELF_ADDR name_offset = 0;
+            if (iter.comment != NULL && iter.comment[0] != '\0') {
+                name_offset = (ELF_ADDR)strhash_hash_lookup(GLOBAL_DCONTEXT,
+                                                            string_htable, iter.comment);
+            }
+            if (!write_shdr(elf_file, name_offset, SHT_PROGBITS, flags,
+                            (ELF_ADDR)iter.vm_start, file_offset,
+                            iter.vm_end - iter.vm_start, /*link=*/0,
+                            /*info=*/0, /*alignment=*/sizeof(ELF_WORD),
+                            /*entry_size=*/0)) {
+                os_close(elf_file);
+                return false;
+            }
+            file_offset += iter.vm_end - iter.vm_start;
+        }
+        memquery_iterator_stop(&iter);
+        // Write the section names section.
+        if (!write_shdr(elf_file, string_table_offset - strlen(".shstrtab"), SHT_STRTAB,
+                        /*flags=*/0, /*virtual_address=*/0, file_offset,
+                        /*section_size=*/string_table_offset, /*link=*/0,
+                        /*info=*/0, /*alignment=*/1,
+                        /*entry_size=*/0)) {
+            os_close(elf_file);
+            return false;
+        }
+    }
+    os_close(elf_file);
+    strhash_hash_destroy(GLOBAL_DCONTEXT, string_htable);
+    return true;
+}
+
+/*
+ * Returns true if a core dump file is written, false otherwise.
+ */
+bool
+os_dump_core_live(void)
+{
+    static thread_id_t current_dumping_thread_id VAR_IN_SECTION(NEVER_PROTECTED_SECTION) =
+        0;
+    thread_id_t current_id = d_r_get_thread_id();
+#ifdef DEADLOCK_AVOIDANCE
+    dcontext_t *dcontext = get_thread_private_dcontext();
+    thread_locks_t *old_thread_owned_locks = NULL;
+#endif
+
+    if (current_id == current_dumping_thread_id) {
+        return false; /* avoid infinite loop */
+    }
+
+#ifdef DEADLOCK_AVOIDANCE
+    /* first turn off deadlock avoidance for this thread (needed for live dump
+     * to try to grab all_threads and thread_initexit locks) */
```

Comment:
But I don't see code grabbing those locks?

Reply:
I copied the DEADLOCK_AVOIDANCE part of the code from win32/os.c. That file has multiple paths to get to os_dump_core_internal(). Do we need the same deadlock mechanism here?

Reply:
The comment is about the all_threads and thread_initexit_locks which you weren't acquiring.  Maybe you should be since this is a global memory snapshot and you'd prefer to suspend the world: suspending all other threads does seem like the right thing to do here.  Is that what the windows code does?

Reply:
I've updated the code to use dr_suspend_all_other_threads_ex() instead of a lock.

win32/os.c uses dump_core_lock to prevent multiple threads from dumping core at the same time, it doesn't suspend other threads.

Reply:
I've updated to code to use synch_with_all_threads instead based comment about using synch_with_all_threads instead.